### DISCUSSION TITLE
Retrospective: lack of clarity on removal of `no-console` rule from `recommended` set

### DISCUSSION THREAD

**Discussion by bjester** [eslint/eslint#16971] [General] [answered]
First off, I realize we're approaching 4 years since the actual change took place to remove the `no-console` rule from the `eslint:recommended` set, but I wanted to offer some thoughts for future improvement.

The [Kolibri development team](https://github.com/learningequality/kolibri) only recently realized that `no-console` was missing. Tracking down where it went missing proved more difficult than expected. The amount of time since the change certainly plays a factor since we upgraded to eslint v6 3 years ago. Although, looking back, it was still difficult to uncover this for the following reasons:
- neither the removal of `no-console` nor changes to `eslint:recommended` were listed in the [final v6.0.0 changelogs](https://eslint.org/blog/2019/06/eslint-v6.0.0-released/)
- the [changelog commit](https://github.com/eslint/eslint/commit/0fc8e62818bc8d0a0a804b59c6110818844df5f3) associated with the change in the changelog does not actually remove `no-console` from the `eslint-recommended.js` because [that file](https://github.com/eslint/eslint/blob/0fc8e62818bc8d0a0a804b59c6110818844df5f3/conf/eslint-recommended.js) was previously refactored to use metadata intended for the docs pages
- the specific change is only listed in the [6.0.0-alpha1 release](https://eslint.org/blog/2019/05/eslint-v6.0.0-alpha.1-released/), but the link to the migration guide, which shares more explicit guidance, is broken
- with the [PR implementing the breaking change](https://github.com/eslint/eslint/pull/11518), there was [indeed confusion](https://github.com/eslint/eslint/pull/11518#discussion_r266277359) about the code change and if it was a breaking change; those conversations are resolved and hidden from view by default and neither the PR nor the referenced issue explicitly mention that `no-console` was _removed_ (and [more confusion later](https://github.com/eslint/eslint/issues/11267))

The only real clarifying notice of the breaking change was in the [migration guide](https://eslint.org/docs/latest/use/migrating-to-6.0.0#eslint-recommended-changes), which is odd considering you would expect such an important breaking change to be clearly outlined in a changelog, and even moreso in the final release changelog. Overall, the whole situation smells.

My suggestions:
- ensure changes during release alphas, betas, and release candidates get included in the final release changelog announcements
- be more explicit with breaking changes in changelogs, and issue and PR descriptions

**Comment by sam3k** [comment, marked as answer]
Hi @bjester, the team will review your suggestions. Thank you for your feedback.

**Comment by nzakas** [comment]
@bjester thanks for the feedback. We have found that changelogs are not a good way to communicate changes such as these, which is why we create Migration Guides for every major release outlining the breaking changes and what to do to address them. Additionally, we've started doing "What's coming" blog posts ([example](https://eslint.org/blog/2021/06/whats-coming-in-eslint-8.0.0/)) to give people a heads up of breaking changes before a major release.

As you point out, v6.0.0 was released four years ago, and our processes and team have changed since then. I'd like to think we are doing a better job now than we did, so hopefully breaking changes going forward will be easier to identify.

**Comment by bjester** [comment]
Thanks for the reply @nzakas. I do really appreciate the in-depth migration guides, and agree that something like that is needed for breaking changes. Although, one aspect to consider is that projects using `eslint` may not have the bandwidth to read migration guides or blogs in order to stay up-to-date, _unless they know they might need to_. For instance, if dependabot opened a PR to update `eslint`, or any other dependency for that matter (which a project likely has many), it would be really helpful have those breadcrumbs in the changelogs and release notes so developers can follow them when changes necessitate that. 

[Here's a recent example](https://github.com/learningequality/kolibri/pull/9999) of a dependabot PR upgrading `eslint`. Unfortunately, we're still behind in versions, so it's encompassing a lot of updates, including a major version upgrade. Obviously, there are ways dependabot could be improved too.

**Reply by nzakas** [reply]
> Although, one aspect to consider is that projects using eslint may not have the bandwidth to read migration guides or blogs in order to stay up-to-date, unless they know they might need to.

I don't think this a problem we can reasonably be expected to solve for. A major version upgrade means, above all else, there are breaking changes and you should probably investigate before upgrading.

I'm not sure there's much else we can do with regards to changelogs, short of manually editing them, which is a big time sink. Every changelog generation tool I've looked at pretty much outputs the same information. If you're aware of one that has a nicer output that you think is more helpful, I'd be happy to take a look.

Another thing to keep in mind: at the moment, we only really have two people working in their spare time doing the majority of the work on this project, so if this is something you care about, the best way to affect change is to help out. :) There's really barely enough bandwidth to keep up with all of the incoming requests, so requests for changes to things that are "good enough" tend to be pushed down in priority.

**Comment by mdjermanovic** [comment]
> * ensure changes during release alphas, betas, and release candidates get included in the final release changelog announcements

I think we already implemented this in https://github.com/eslint/eslint-release/pull/37.

[v8.0.0 release blog post](https://eslint.org/blog/2021/10/eslint-v8.0.0-released/), [v8.0.0 GitHub release](https://github.com/eslint/eslint/releases/tag/v8.0.0), and v8.0.0 entry in [CHANGELOG.md](https://github.com/eslint/eslint/blob/main/CHANGELOG.md) include changes from all prerelease versions.

The same for v7.0.0.

v6.0.0 indeed wasn't including all changes,.

**Comment by sam3k** [comment]
> Although, one aspect to consider is that projects using eslint may not have the bandwidth to read migration guides or blogs in order to stay up-to-date, unless they know they might need to.

Imho, when doing an upgrade that falls into "major" such as in this case, it should be expected to have breaking changes and operate under the assumption that things will break or to have been deprecated.

**Comment by bmish** [comment]
In addition to mentioning such breaking changes in the migration guide, I believe we should include the names of all the rules being added or removed from the `recommended` config in the commit message itself, so that this key information shows up in the changelog. That's exactly what I do for major versions of plugins I maintain, for example:
* https://github.com/eslint-community/eslint-plugin-eslint-plugin/releases/tag/v5.0.0
   > Breaking: Add `no-assert-equal` to recommended config
* https://github.com/platinumazure/eslint-plugin-qunit/releases/tag/v7.0.0
   > add `recommended` rules `prefer-message-ids`, `prefer-output-null`, `no-missing-message-ids`, `no-unused-message-ids`
